src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,309 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import category_theory.epi_mono
+import group_theory.quotient_group
+import algebra.category.Group.basic
+
+/-!
+# Monomorphisms and epimorphisms in `Group`
+In this file, we prove monomorphisms in category of group are injective homomorphisms and
+epimorphisms are surjective homomorphisms.
+-/
+
+noncomputable theory
+
+universes u v
+
+namespace monoid_hom
+
+open quotient_group
+
+variables {A : Type u} {B : Type v}
+
+variables [group A] [group B]
+
+@[to_additive add_monoid_hom.ker_eq_bot_of_cancel]
+lemma ker_eq_bot_of_cancel {f : A →* B}
+  (h : ∀ (u v : f.ker →* A), f.comp u = f.comp v → u = v) :
+  f.ker = (⊥ : subgroup A) :=
+begin
+  specialize h 1 f.ker.subtype begin
+    ext1,
+    simp only [comp_one, one_apply, coe_comp, subgroup.coe_subtype, function.comp_app],
+    rw [←subtype.val_eq_coe, f.mem_ker.mp x.2],
+  end,
+  rw [←subgroup.subtype_range f.ker, ←h, range_one],
+end
```

Comment:
```suggestion
@[to_additive add_monoid_hom.ker_eq_bot_of_cancel]
lemma ker_eq_bot_of_cancel {f : A →* B} (h : ∀ (u v : f.ker →* A), f.comp u = f.comp v → u = v) :
  f.ker = (⊥ : subgroup A) :=
by simpa using _root_.congr_arg range (h f.ker.subtype 1 (by tidy))
```
You don't have to golf that hard if you don't want to, but please at least use a subgoal instead of the inner `begin..end` block.